flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkStateSnapshotController.java

```diff
@@ -0,0 +1,207 @@
+import io.javaoperatorsdk.operator.api.reconciler.ErrorStatusUpdateControl;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceContext;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceInitializer;
+import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
+import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
+import io.javaoperatorsdk.operator.processing.event.source.EventSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+
+/** Controller that runs the main reconcile loop for {@link FlinkStateSnapshot}. */
+@ControllerConfiguration
+public class FlinkStateSnapshotController
+        implements Reconciler<FlinkStateSnapshot>,
+                ErrorStatusHandler<FlinkStateSnapshot>,
+                EventSourceInitializer<FlinkStateSnapshot>,
+                Cleaner<FlinkStateSnapshot> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkStateSnapshotController.class);
+
+    private final Set<FlinkResourceValidator> validators;
+    private final FlinkResourceContextFactory ctxFactory;
+    private final StateSnapshotReconciler reconciler;
+    private final StateSnapshotObserver observer;
+    private final EventRecorder eventRecorder;
+    private final StatusRecorder<FlinkStateSnapshot, FlinkStateSnapshotStatus> statusRecorder;
+
+    public FlinkStateSnapshotController(
+            Set<FlinkResourceValidator> validators,
+            FlinkResourceContextFactory ctxFactory,
+            StateSnapshotReconciler reconciler,
+            StateSnapshotObserver observer,
+            EventRecorder eventRecorder,
+            StatusRecorder<FlinkStateSnapshot, FlinkStateSnapshotStatus> statusRecorder) {
+        this.validators = validators;
+        this.ctxFactory = ctxFactory;
+        this.reconciler = reconciler;
+        this.observer = observer;
+        this.eventRecorder = eventRecorder;
+        this.statusRecorder = statusRecorder;
+    }
+
+    @Override
+    public UpdateControl<FlinkStateSnapshot> reconcile(
+            FlinkStateSnapshot flinkStateSnapshot, Context<FlinkStateSnapshot> josdkContext) {
+        statusRecorder.updateStatusFromCache(flinkStateSnapshot);
+
+        var ctx = ctxFactory.getFlinkStateSnapshotContext(flinkStateSnapshot, josdkContext);
+
+        // observe
+        observer.observe(ctx);
+
+        // validate
+        if (!validateSavepoint(ctx)) {
+            statusRecorder.patchAndCacheStatus(flinkStateSnapshot, ctx.getKubernetesClient());
+            UpdateControl<FlinkStateSnapshot> updateControl = UpdateControl.noUpdate();
+            return updateControl.rescheduleAfter(
+                    ctx.getOperatorConfig().getReconcileInterval().toMillis());
+        }
+
+        // reconcile
+        try {
+            statusRecorder.patchAndCacheStatus(flinkStateSnapshot, ctx.getKubernetesClient());
```

Comment:
I think for FlinkStateSnapshot resources we should avoid using the custom status patching. We can just rely on the JOSDK built in Update control mechanism. 

We are using a custom patching logic for the deployments because we are executing destructive actions before which we must record the status in some cases. Here I believe it's not the case in general.

Reply:
StatusRecorder will also notify `FlinkResourceListener` instances, I am not sure if we should give up that functionality for the snapshot resource. But I agree that we should use the JOSDK update mechanism if possible, I will try to refactor it.

Reply:
Done in commit 0315d1062dcf0b81e61af77be4a0c12e9330398b - I added some logic to decide if the status was modified, and notify listeners and `patchStatus` if it changed.